### PR TITLE
add download logs endpoint

### LINKED ISSUE #369: New monitoring endpoint: provide all logs
https://github.com/fractal-analytics-platform/fractal-server/issues/52#issuecomment-1359409604 covers the essential logging & user feedback by giving a way to see the error messages when something failed. A future enhancement would be to give users access to the logs for each task, e.g. to see if there were any warnings for a given task. I think the core thing here would be access to the .err files.

This is not urgent though. Will also come up as a discussion around the web frontend as well :)

### PR DESCRIPTION
**Pull request by mfranzon**
Close #369

### CODE REVIEW

File: fractal_server/app/api/v1/job.py [hunk 1]
```diff
@@ -46,3 +47,46 @@ async def get_job(
...
 
     return job_read
+
+
+@router.get("/download/{job_id}", response_class=StreamingResponse)
+async def download_job_logs(
+    job_id: int,
+    user: User = Depends(current_active_user),
+    db: AsyncSession = Depends(get_db),
+) -> Optional[ApplyWorkflow]:
```
Comment (tcompa) [inline fractal_server/app/api/v1/job.py:57]: typo

File: fractal_server/app/api/v1/job.py [hunk 2]
```diff
@@ -46,3 +47,46 @@ async def get_job(
...
+        project_id=job.project_id, user_id=user.id, db=db
+    )
+
+    job_read = ApplyWorkflowRead(**job.dict())
+
+    from devtools import debug
+
+    debug(job_read)
+
+    from io import BytesIO
```
Comment (tcompa) [inline fractal_server/app/api/v1/job.py:73]: Move imports

File: fractal_server/app/api/v1/job.py [hunk 3]
```diff
@@ -46,3 +47,46 @@ async def get_job(
...
+    from zipfile import ZipFile
+    from zipfile import ZIP_DEFLATED
+    import os
+
+    io = BytesIO()
+    zip_sub_dir = (
+        f"{job_read.project_id}_{job_read.workflow_id}_{job_id}_archive"
```
Comment (tcompa) [inline fractal_server/app/api/v1/job.py:80]: Use `job_read.working_dir.name`

File: fractal_server/app/api/v1/job.py [hunk 4]
```diff
@@ -46,3 +47,46 @@ async def get_job(
...
+    )
+    zip_filename = "%s.zip" % zip_sub_dir
+    with ZipFile(io, mode="w", compression=ZIP_DEFLATED) as zip:
+        for fpath in os.listdir(job_read.working_dir):
```
Comment (tcompa) [inline fractal_server/app/api/v1/job.py:84]: `job_read.working_dir.glob()`

File: fractal_server/app/api/v1/job.py [hunk 5]
```diff
@@ -46,3 +47,46 @@ async def get_job(
...
+            zip.write(job_read.working_dir / fpath)
+        # close zip
+        zip.close()
```
Comment (tcompa) [inline fractal_server/app/api/v1/job.py:87]: Already taken care of by `with Zipfile() as zip`?

Comment (tcompa) [reply fractal_server/app/api/v1/job.py:87]: I take it back.
But then let's use an instance rather than `with`

Comment (mfranzon) [reply fractal_server/app/api/v1/job.py:87]: > ZipFile is also a context manager and therefore supports the [with](https://docs.python.org/3/reference/compound_stmts.html#with) statement. ...closed after the with statement’s suite is finished—even if an exception occurs...

my bad, it's already a context manager. Just remove .close() method

File: fractal_server/app/api/v1/job.py (hunk 3 above)
Comment (mfranzon) [reply fractal_server/app/api/v1/job.py:80]: I'm using `job.dict()["working_dir"] `so `.name` it is not available for strings.

**Pull request merged**